File: internal/domain/editable_parser_test.go
package domain

import (
	"testing"
)

func TestParseLine_SymbolTypes(t *testing.T) {
	parser := NewEditableDocumentParser()

	tests := []struct {
		name     string
		input    string
		wantType EntryType
		wantText string
	}{
		{
			name:     "task with dot",
			input:    ". Buy groceries",
			wantType: EntryTypeTask,
			wantText: "Buy groceries",
		},
		{
			name:     "note with dash",
			input:    "- Meeting went well",
			wantType: EntryTypeNote,
			wantText: "Meeting went well",
		},
		{
			name:     "event with o",
			input:    "o Team standup at 10am",
			wantType: EntryTypeEvent,
			wantText: "Team standup at 10am",
		},
		{
			name:     "done with x",
			input:    "x Finished report",
			wantType: EntryTypeDone,
			wantText: "Finished report",
		},
		{
			name:     "cancelled with tilde",
			input:    "~ No longer needed",
			wantType: EntryTypeCancelled,
			wantText: "No longer needed",
		},
		{
			name:     "question with question mark",
			input:    "? How does auth work",
			wantType: EntryTypeQuestion,
			wantText: "How does auth work",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			line := parser.ParseLine(tt.input, 1)

			if line.Symbol != tt.wantType {
				t.Errorf("got symbol %v, want %v", line.Symbol, tt.wantType)
			}
			if line.Content != tt.wantText {
				t.Errorf("got content %q, want %q", line.Content, tt.wantText)
			}
			if !line.IsValid {
				t.Errorf("expected line to be valid, got error: %s", line.ErrorMessage)
			}
		})
	}
}

func TestParseLine_Priority(t *testing.T) {
	parser := NewEditableDocumentParser()

	tests := []struct {
		name         string
		input        string
		wantPriority Priority
		wantContent  string
	}{
		{
			name:         "high priority with triple bang",
			input:        ". !!! Urgent task",
			wantPriority: PriorityHigh,
			wantContent:  "Urgent task",
		},
		{
			name:         "medium priority with double bang",
			input:        ". !! Important task",
			wantPriority: PriorityMedium,
			wantContent:  "Important task",
		},
		{
			name:         "low priority with single bang",
			input:        ". ! Minor task",
			wantPriority: PriorityLow,
			wantContent:  "Minor task",
		},
		{
			name:         "no priority",
			input:        ". Normal task",
			wantPriority: PriorityNone,
			wantContent:  "Normal task",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			line := parser.ParseLine(tt.input, 1)

			if line.Priority != tt.wantPriority {
				t.Errorf("got priority %v, want %v", line.Priority, tt.wantPriority)
			}
			if line.Content != tt.wantContent {
				t.Errorf("got content %q, want %q", line.Content, tt.wantContent)
			}
		})
	}
}

func TestParseLine_Indentation(t *testing.T) {
	parser := NewEditableDocumentParser()

	tests := []struct {
		name      string
		input     string
		wantDepth int
	}{
		{
			name:      "no indentation",
			input:     ". Root task",
			wantDepth: 0,
		},
		{
			name:      "one level (2 spaces)",
			input:     "  . Child task",
			wantDepth: 1,
		},
		{
			name:      "two levels (4 spaces)",
			input:     "    . Grandchild task",
			wantDepth: 2,
		},
		{
			name:      "three levels (6 spaces)",
			input:     "      . Great-grandchild",
			wantDepth: 3,
		},
		{
			name:      "tab indentation normalized",
			input:     "\t. Tab indented",
			wantDepth: 1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			line := parser.ParseLine(tt.input, 1)

			if line.Depth != tt.wantDepth {
				t.Errorf("got depth %d, want %d", line.Depth, tt.wantDepth)
			}
		})
	}
}

func TestParseLine_InvalidLines(t *testing.T) {
	parser := NewEditableDocumentParser()

	tests := []struct {
		name         string
		input        string
		wantValid    bool
		wantErrorMsg string
	}{
		{
			name:         "unknown symbol",
			input:        "^ Invalid entry",
			wantValid:    false,
			wantErrorMsg: "Unknown entry type",
		},
		{
			name:         "missing content",
			input:        ".",
			wantValid:    false,
			wantErrorMsg: "Entry content required",
		},
		{
			name:         "only whitespace content",
			input:        ".   ",
			wantValid:    false,
			wantErrorMsg: "Entry content required",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			line := parser.ParseLine(tt.input, 1)

			if line.IsValid != tt.wantValid {
				t.Errorf("got IsValid=%v, want %v", line.IsValid, tt.wantValid)
			}
			if tt.wantErrorMsg != "" && line.ErrorMessage != tt.wantErrorMsg {
				t.Errorf("got error %q, want %q", line.ErrorMessage, tt.wantErrorMsg)
			}
		})
	}
}

func TestParseLine_HeaderLines(t *testing.T) {
	parser := NewEditableDocumentParser()

	tests := []struct {
		name       string
		input      string
		wantHeader bool
	}{
		{
			name:       "header line with dashes",
			input:      "── Monday, Jan 27 ──────────────────",
			wantHeader: true,
		},
		{
			name:       "regular entry",
			input:      ". Not a header",
			wantHeader: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			line := parser.ParseLine(tt.input, 1)

			if line.IsHeader != tt.wantHeader {
				t.Errorf("got IsHeader=%v, want %v", line.IsHeader, tt.wantHeader)
			}
		})
	}
}

func TestParse_FullDocument(t *testing.T) {
	parser := NewEditableDocumentParser()

	input := `── Monday, Jan 27 ──────────────────
. Buy groceries
  . Milk
  . Eggs
- Meeting went well
. !! Urgent: fix prod bug
x Deployed hotfix`

	doc, err := parser.Parse(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(doc.Lines) != 7 {
		t.Fatalf("expected 7 lines, got %d", len(doc.Lines))
	}

	if !doc.Lines[0].IsHeader {
		t.Error("expected first line to be header")
	}

	if doc.Lines[1].Content != "Buy groceries" || doc.Lines[1].Depth != 0 {
		t.Errorf("line 1: got content=%q depth=%d, want content='Buy groceries' depth=0",
			doc.Lines[1].Content, doc.Lines[1].Depth)
	}

	if doc.Lines[2].Content != "Milk" || doc.Lines[2].Depth != 1 {
		t.Errorf("line 2: got content=%q depth=%d, want content='Milk' depth=1",
			doc.Lines[2].Content, doc.Lines[2].Depth)
	}

	if doc.Lines[5].Priority != PriorityMedium {
		t.Errorf("line 5: got priority %v, want PriorityMedium", doc.Lines[5].Priority)
	}

	if doc.Lines[6].Symbol != EntryTypeDone {
		t.Errorf("line 6: got symbol %v, want EntryTypeDone", doc.Lines[6].Symbol)
	}
}

func TestParse_EmptyLines(t *testing.T) {
	parser := NewEditableDocumentParser()

	input := `. Task one

. Task two`

	doc, err := parser.Parse(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	validLines := 0
	for _, line := range doc.Lines {
		if line.IsValid && !line.IsHeader {
			validLines++
		}
	}

	if validLines != 2 {
		t.Errorf("expected 2 valid entry lines, got %d", validLines)
	}
}

func TestSerialize_BasicEntries(t *testing.T) {
	tests := []struct {
		name    string
		entries []Entry
		want    string
	}{
		{
			name: "task entry",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Buy groceries", Depth: 0},
			},
			want: ". Buy groceries",
		},
		{
			name: "note entry",
			entries: []Entry{
				{Type: EntryTypeNote, Content: "Meeting went well", Depth: 0},
			},
			want: "- Meeting went well",
		},
		{
			name: "event entry",
			entries: []Entry{
				{Type: EntryTypeEvent, Content: "Team standup at 10am", Depth: 0},
			},
			want: "o Team standup at 10am",
		},
		{
			name: "done entry",
			entries: []Entry{
				{Type: EntryTypeDone, Content: "Finished report", Depth: 0},
			},
			want: "x Finished report",
		},
		{
			name: "cancelled entry",
			entries: []Entry{
				{Type: EntryTypeCancelled, Content: "No longer needed", Depth: 0},
			},
			want: "~ No longer needed",
		},
		{
			name: "question entry",
			entries: []Entry{
				{Type: EntryTypeQuestion, Content: "How does auth work", Depth: 0},
			},
			want: "? How does auth work",
		},
		{
			name: "migrated entry",
			entries: []Entry{
				{Type: EntryTypeMigrated, Content: "Moved to next week", Depth: 0},
			},
			want: "> Moved to next week",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Serialize(tt.entries)
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestSerialize_WithPriority(t *testing.T) {
	tests := []struct {
		name    string
		entries []Entry
		want    string
	}{
		{
			name: "high priority",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Urgent task", Priority: PriorityHigh, Depth: 0},
			},
			want: ". !!! Urgent task",
		},
		{
			name: "medium priority",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Important task", Priority: PriorityMedium, Depth: 0},
			},
			want: ". !! Important task",
		},
		{
			name: "low priority",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Minor task", Priority: PriorityLow, Depth: 0},
			},
			want: ". ! Minor task",
		},
		{
			name: "no priority",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Normal task", Priority: PriorityNone, Depth: 0},
			},
			want: ". Normal task",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Serialize(tt.entries)
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestSerialize_WithIndentation(t *testing.T) {
	tests := []struct {
		name    string
		entries []Entry
		want    string
	}{
		{
			name: "depth 1",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Child task", Depth: 1},
			},
			want: "  . Child task",
		},
		{
			name: "depth 2",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Grandchild task", Depth: 2},
			},
			want: "    . Grandchild task",
		},
		{
			name: "depth 3",
			entries: []Entry{
				{Type: EntryTypeTask, Content: "Great-grandchild", Depth: 3},
			},
			want: "      . Great-grandchild",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Serialize(tt.entries)
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestSerialize_MultipleEntries(t *testing.T) {
	entries := []Entry{
		{Type: EntryTypeTask, Content: "Buy groceries", Depth: 0},
		{Type: EntryTypeTask, Content: "Milk", Depth: 1},
		{Type: EntryTypeTask, Content: "Eggs", Depth: 1},
		{Type: EntryTypeNote, Content: "Meeting went well", Depth: 0},
		{Type: EntryTypeTask, Content: "Urgent fix", Priority: PriorityMedium, Depth: 0},
		{Type: EntryTypeDone, Content: "Deployed hotfix", Depth: 0},
	}

	want := `. Buy groceries
  . Milk
  . Eggs
- Meeting went well
. !! Urgent fix
x Deployed hotfix`

	got := Serialize(entries)
	if got != want {
		t.Errorf("got:\n%s\n\nwant:\n%s", got, want)
	}
}

func TestSerialize_BuildsHierarchyFromParentID(t *testing.T) {
	parentID := int64(1)
	childID := int64(2)

	entries := []Entry{
		{ID: 1, Type: EntryTypeNote, Content: "note 5", Depth: 0},
		{ID: 2, Type: EntryTypeEvent, Content: "event 1", Depth: 0},
		{ID: 3, Type: EntryTypeNote, Content: "note 1x", Depth: 1, ParentID: &childID},
		{ID: 4, Type: EntryTypeNote, Content: "note 6x", Depth: 1, ParentID: &parentID},
	}

	got := Serialize(entries)
	want := `- note 5
  - note 6x
o event 1
  - note 1x`

	if got != want {
		t.Errorf("got:\n%s\n\nwant:\n%s", got, want)
	}
}

func TestSerialize_NestedHierarchyFromParentID(t *testing.T) {
	parentID := int64(10)
	childID := int64(20)

	entries := []Entry{
		{ID: 10, Type: EntryTypeTask, Content: "parent", Depth: 0},
		{ID: 20, Type: EntryTypeNote, Content: "child", Depth: 1, ParentID: &parentID},
		{ID: 30, Type: EntryTypeNote, Content: "grandchild", Depth: 2, ParentID: &childID},
	}

	got := Serialize(entries)
	want := `. parent
  - child
    - grandchild`

	if got != want {
		t.Errorf("got:\n%s\n\nwant:\n%s", got, want)
	}
}

func TestSerialize_EmptyEntries(t *testing.T) {
	got := Serialize(nil)
	if got != "" {
		t.Errorf("expected empty string for nil entries, got %q", got)
	}

	got = Serialize([]Entry{})
	if got != "" {
		t.Errorf("expected empty string for empty entries, got %q", got)
	}
}

File: frontend/src/lib/editableParser.test.ts
import { describe, it, expect } from 'vitest'
import { parseLine, parseDocument } from './editableParser'

describe('parseLine', () => {
  describe('symbol recognition', () => {
    it('parses task symbol (.)', () => {
      const result = parseLine('. Buy groceries', 1)

      expect(result.symbol).toBe('.')
      expect(result.content).toBe('Buy groceries')
      expect(result.isValid).toBe(true)
    })

    it('parses note symbol (-)', () => {
      const result = parseLine('- Meeting went well', 1)

      expect(result.symbol).toBe('-')
      expect(result.content).toBe('Meeting went well')
      expect(result.isValid).toBe(true)
    })

    it('parses event symbol (o)', () => {
      const result = parseLine('o Team standup at 10am', 1)

      expect(result.symbol).toBe('o')
      expect(result.content).toBe('Team standup at 10am')
      expect(result.isValid).toBe(true)
    })

    it('parses done symbol (x)', () => {
      const result = parseLine('x Finished report', 1)

      expect(result.symbol).toBe('x')
      expect(result.content).toBe('Finished report')
      expect(result.isValid).toBe(true)
    })

    it('parses cancelled symbol (~)', () => {
      const result = parseLine('~ No longer needed', 1)

      expect(result.symbol).toBe('~')
      expect(result.content).toBe('No longer needed')
      expect(result.isValid).toBe(true)
    })

    it('parses question symbol (?)', () => {
      const result = parseLine('? How does auth work', 1)

      expect(result.symbol).toBe('?')
      expect(result.content).toBe('How does auth work')
      expect(result.isValid).toBe(true)
    })

    it('parses migrated symbol (>)', () => {
      const result = parseLine('> Moved task', 1)

      expect(result.symbol).toBe('>')
      expect(result.content).toBe('Moved task')
      expect(result.isValid).toBe(true)
    })

    it('marks unknown symbol as invalid', () => {
      const result = parseLine('^ Unknown', 1)

      expect(result.isValid).toBe(false)
      expect(result.errorMessage).toBe('Unknown entry type')
    })
  })

  describe('priority parsing', () => {
    it('parses high priority (!!!)', () => {
      const result = parseLine('. !!! Urgent task', 1)

      expect(result.symbol).toBe('.')
      expect(result.priority).toBe(1)
      expect(result.content).toBe('Urgent task')
    })

    it('parses medium priority (!!)', () => {
      const result = parseLine('. !! High priority', 1)

      expect(result.symbol).toBe('.')
      expect(result.priority).toBe(2)
      expect(result.content).toBe('High priority')
    })

    it('parses low priority (!)', () => {
      const result = parseLine('. ! Low priority', 1)

      expect(result.symbol).toBe('.')
      expect(result.priority).toBe(3)
      expect(result.content).toBe('Low priority')
    })

    it('defaults to no priority (0) when no markers', () => {
      const result = parseLine('. Normal task', 1)

      expect(result.priority).toBe(0)
      expect(result.content).toBe('Normal task')
    })
  })

  describe('indentation', () => {
    it('parses depth 0 (no indentation)', () => {
      const result = parseLine('. Root task', 1)

      expect(result.depth).toBe(0)
    })

    it('parses depth 1 (2 spaces)', () => {
      const result = parseLine('  . Child task', 1)

      expect(result.depth).toBe(1)
    })

    it('parses depth 2 (4 spaces)', () => {
      const result = parseLine('    . Grandchild', 1)

      expect(result.depth).toBe(2)
    })

    it('normalizes tab to spaces', () => {
      const result = parseLine('\t. Tabbed task', 1)

      expect(result.depth).toBe(1)
    })

    it('rounds odd indentation to nearest level', () => {
      const result = parseLine('   . Odd indent', 1)

      expect(result.depth).toBe(2) // 3 spaces rounds to 2 (depth 1)
    })
  })

  describe('migration syntax', () => {
    it('detects migration pattern with ISO date', () => {
      const result = parseLine('>[2026-01-29] Call dentist', 1)

      expect(result.symbol).toBe('>')
      expect(result.migrationTarget).toBe('2026-01-29')
      expect(result.content).toBe('Call dentist')
      expect(result.isValid).toBe(true)
    })

    it('detects migration pattern with natural language', () => {
      const result = parseLine('>[tomorrow] Review PR', 1)

      expect(result.symbol).toBe('>')
      expect(result.migrationTarget).toBe('tomorrow')
      expect(result.content).toBe('Review PR')
    })

    it('detects migration pattern with multi-word date', () => {
      const result = parseLine('>[next monday] Submit report', 1)

      expect(result.symbol).toBe('>')
      expect(result.migrationTarget).toBe('next monday')
      expect(result.content).toBe('Submit report')
    })

    it('handles indented migration', () => {
      const result = parseLine('  >[tomorrow] Indented migration', 1)

      expect(result.depth).toBe(1)
      expect(result.symbol).toBe('>')
      expect(result.migrationTarget).toBe('tomorrow')
      expect(result.content).toBe('Indented migration')
    })
  })

  describe('edge cases', () => {
    it('returns empty result for empty line', () => {
      const result = parseLine('', 1)

      expect(result.isEmpty).toBe(true)
      expect(result.isValid).toBe(true)
    })

    it('returns empty result for whitespace-only line', () => {
      const result = parseLine('   ', 1)

      expect(result.isEmpty).toBe(true)
      expect(result.isValid).toBe(true)
    })

    it('skips header lines (──)', () => {
      const result = parseLine('── Monday, Jan 27 ──────────────────', 1)

      expect(result.isHeader).toBe(true)
      expect(result.isValid).toBe(true)
    })

    it('marks missing content as error', () => {
      const result = parseLine('. ', 1)

      expect(result.isValid).toBe(false)
      expect(result.errorMessage).toBe('Entry content required')
    })

    it('preserves line number in result', () => {
      const result = parseLine('. Test', 42)

      expect(result.lineNumber).toBe(42)
    })

    it('handles symbol-only line as missing content', () => {
      const result = parseLine('.', 1)

      expect(result.isValid).toBe(false)
      expect(result.errorMessage).toBe('Entry content required')
    })
  })
})

describe('parseDocument', () => {
  it('parses multiple lines', () => {
    const doc = `. Buy groceries
- Meeting went well
x Finished report`

    const result = parseDocument(doc)

    expect(result.lines).toHaveLength(3)
    expect(result.lines[0].symbol).toBe('.')
    expect(result.lines[1].symbol).toBe('-')
    expect(result.lines[2].symbol).toBe('x')
  })

  it('assigns sequential line numbers', () => {
    const doc = `. First
. Second
. Third`

    const result = parseDocument(doc)

    expect(result.lines[0].lineNumber).toBe(1)
    expect(result.lines[1].lineNumber).toBe(2)
    expect(result.lines[2].lineNumber).toBe(3)
  })

  it('handles document with headers', () => {
    const doc = `── Monday, Jan 27 ──────────────────
. Buy groceries
- Meeting went well`

    const result = parseDocument(doc)

    expect(result.lines).toHaveLength(3)
    expect(result.lines[0].isHeader).toBe(true)
    expect(result.lines[1].symbol).toBe('.')
  })

  it('handles empty document', () => {
    const result = parseDocument('')

    expect(result.lines).toHaveLength(0)
    expect(result.isValid).toBe(true)
    expect(result.errors).toHaveLength(0)
  })

  it('collects errors from invalid lines', () => {
    const doc = `. Valid task
^ Invalid line
. Another valid`

    const result = parseDocument(doc)

    expect(result.isValid).toBe(false)
    expect(result.errors).toHaveLength(1)
    expect(result.errors[0].lineNumber).toBe(2)
    expect(result.errors[0].message).toBe('Unknown entry type')
  })

  it('handles hierarchical document', () => {
    const doc = `. Parent task
  . Child task
    . Grandchild
  - Sibling note`

    const result = parseDocument(doc)

    expect(result.lines[0].depth).toBe(0)
    expect(result.lines[1].depth).toBe(1)
    expect(result.lines[2].depth).toBe(2)
    expect(result.lines[3].depth).toBe(1)
  })

  it('preserves empty lines', () => {
    const doc = `. First

. Second`

    const result = parseDocument(doc)

    expect(result.lines).toHaveLength(3)
    expect(result.lines[1].isEmpty).toBe(true)
  })
})
